Add WK API to be notified of background fetch changes

youennf opened the pull request:
#### 6de9a281e1ebf08e353a6f6184b51edea9296b3f
<pre>
Add WK API to be notified of background fetch changes
<a href="https://bugs.webkit.org/show_bug.cgi?id=253305">https://bugs.webkit.org/show_bug.cgi?id=253305</a>
rdar://problem/106190496

Reviewed by Chris Dumez.

Add a Website data store delegate to be notified of new background fetch.
Add a Website data store private method to get backgorund fetch state information from a background fetch identifier.
The changes are: Addition, Removal, Update.
The states are exposed as a NSDictionary.
We add new test runner methods to make use of these website data store APIs.

Drive-by fixes to make the records no longer available once the termina background fetch (success/error/abort) event is fully handled.
This is necessary to pass the updated background-fetch layout test.

* LayoutTests/http/wpt/background-fetch/background-fetch-abort.window.js:
(promise_test.async t):
* Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
(WebCore::WorkerSWClientConnection::retrieveRecordResponseBody):
* Source/WebCore/workers/service/background-fetch/BackgroundFetch.cpp:
(WebCore::BackgroundFetch::updateBackgroundFetchStatus):
(WebCore::BackgroundFetch::unsetRecordsAvailableFlag):
* Source/WebCore/workers/service/background-fetch/BackgroundFetch.h:
(WebCore::BackgroundFetch::options const):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.cpp:
(WebCore::BackgroundFetchEngine::startBackgroundFetch):
(WebCore::BackgroundFetchEngine::notifyBackgroundFetchUpdate):
(WebCore::BackgroundFetchEngine::addFetchFromStore):
(WebCore::BackgroundFetchEngine::backgroundFetch const):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.h:
* Source/WebCore/workers/service/background-fetch/BackgroundFetchFailureReason.idl:
* Source/WebCore/workers/service/background-fetch/BackgroundFetchResult.idl:
* Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
(WebCore::ServiceWorkerThread::queueTaskToFireBackgroundFetchEvent):
* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::fireBackgroundFetchEvent):
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::getBackgroundFetchState):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::getBackgroundFetchState):
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreImpl.cpp:
(WebKit::BackgroundFetchStoreImpl::clearFetch):
(WebKit::BackgroundFetchStoreImpl::clearAllFetches):
(WebKit::BackgroundFetchStoreImpl::storeFetch):
(WebKit::BackgroundFetchStoreImpl::storeFetchResponseBodyChunk):
(WebKit::BackgroundFetchStoreImpl::getBackgroundFetchState):
* Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreImpl.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::notifyBackgroundFetchChange):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/Shared/BackgroundFetchChange.h: Added.
* Source/WebKit/Shared/BackgroundFetchState.h: Added.
* Source/WebKit/Shared/BackgroundFetchState.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/BackgroundFetchStateCocoa.mm: Added.
(WebKit::BackgroundFetchState::toDictionary const):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _getBackgroundFetchState:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::getAllBackgroundFetchIdentifiers):
(WebKit::NetworkProcessProxy::getBackgroundFetchState):
(WebKit::NetworkProcessProxy::notifyBackgroundFetchChange):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreClient.h:
(WebKit::WebsiteDataStoreClient::notifyBackgroundFetchChange):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::InjectedBundle::lastAddedBackgroundFetchIdentifier const):
(WTR::InjectedBundle::lastRemovedBackgroundFetchIdentifier const):
(WTR::InjectedBundle::lastUpdatedBackgroundFetchIdentifier const):
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.h:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::lastAddedBackgroundFetchIdentifier const):
(WTR::TestRunner::lastRemovedBackgroundFetchIdentifier const):
(WTR::TestRunner::lastUpdatedBackgroundFetchIdentifier const):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::lastAddedBackgroundFetchIdentifier const):
(WTR::TestController::lastRemovedBackgroundFetchIdentifier const):
(WTR::TestController::lastUpdatedBackgroundFetchIdentifier const):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveSynchronousMessageFromInjectedBundle):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::lastAddedBackgroundFetchIdentifier const):
(WTR::TestController::lastRemovedBackgroundFetchIdentifier const):
(WTR::TestController::lastUpdatedBackgroundFetchIdentifier const):
* Tools/WebKitTestRunner/cocoa/TestWebsiteDataStoreDelegate.h:
* Tools/WebKitTestRunner/cocoa/TestWebsiteDataStoreDelegate.mm:
(-[TestWebsiteDataStoreDelegate requestStorageSpace:frameOrigin:quota:currentSize:spaceRequired:decisionHandler:]):
(-[TestWebsiteDataStoreDelegate websiteDataStore:openWindow:fromServiceWorkerOrigin:completionHandler:]):
(-[TestWebsiteDataStoreDelegate notifyBackgroundFetchChange:change:]):
(-[TestWebsiteDataStoreDelegate lastAddedBackgroundFetchIdentifier]):
(-[TestWebsiteDataStoreDelegate lastRemovedBackgroundFetchIdentifier]):
(-[TestWebsiteDataStoreDelegate lastUpdatedBackgroundFetchIdentifier]):

Canonical link: <a href="https://commits.webkit.org/261261@main">https://commits.webkit.org/261261@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/acc3adaa49a4ad14b8edd869a30b7dac04dae6d8

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/111092 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/20232 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/43656 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/87/builds/2519 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/119936 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/115049 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/21605 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/11334 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/86/builds/2156 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/116847 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/78/builds/16046 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/99230 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/103603 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/98003 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/30894 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/44517 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/12749 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/32228 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/86420 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/13290 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/9215 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/18708 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/51824 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/7812 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/15241 "Built successfully") | | | 
| | | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.h:
```diff
@@ -57,6 +57,7 @@ class BackgroundFetchEngine : public CanMakeWeakPtr<BackgroundFetchEngine> {
 
     void remove(SWServerRegistration&);
 
+    WEBCORE_EXPORT WeakPtr<BackgroundFetch> backgroundFetch(const ServiceWorkerRegistrationKey&, const String&);
```
Can this be const?

youennf replied:
Yes, fixed

cdumez commented on Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreImpl.h:
```diff
@@ -42,13 +42,15 @@ class SWServer;
     ~BackgroundFetchStoreImpl();
 
     void getAllBackgroundFetchIdentifiers(CompletionHandler<void(Vector<String>&&)>&&);
+    void getBackgroundFetchState(const String&, CompletionHandler<void(std::optional<BackgroundFetchState>&&)>&&);
```
Why are we using get prefixes for getters?

youennf replied:
It is an async getter, hence why the get, although other prefixes like retrieve are used as well.

cdumez commented on Source/WebKit/Shared/Cocoa/BackgroundFetchStateCocoa.mm:
```diff
@@ -0,0 +1,58 @@
+
+namespace WebKit {
+
+#define WebKitPushDataKey @"WebKitPushData"
```
Where are these used?

youennf replied:
Removed

cdumez commented on Source/WebKit/Shared/Cocoa/BackgroundFetchStateCocoa.mm:
```diff
@@ -0,0 +1,58 @@
+        @"Uploaded" : @(uploaded),
+        @"Result" : (NSString *)(convertEnumerationToString(result)),
+        @"FailureReason" : (NSString *)(convertEnumerationToString(failureReason)),
+        @"IsActive" : (NSString *)isActive,
```
You're casting a boolean to a NSString* ? You likely want `@(isActive)`

youennf replied:
fixed

youennf committed:
Add WK API to be notified of background fetch changes
https://bugs.webkit.org/show_bug.cgi?id=253305
rdar://problem/106190496

Reviewed by Chris Dumez.

Add a Website data store delegate to be notified of new background fetch.
Add a Website data store private method to get backgorund fetch state information from a background fetch identifier.
The changes are: Addition, Removal, Update.
The states are exposed as a NSDictionary.
We add new test runner methods to make use of these website data store APIs.

Drive-by fixes to make the records no longer available once the termina background fetch (success/error/abort) event is fully handled.
This is necessary to pass the updated background-fetch layout test.

* LayoutTests/http/wpt/background-fetch/background-fetch-abort.window.js:
(promise_test.async t):
* Source/WebCore/workers/service/WorkerSWClientConnection.cpp:
(WebCore::WorkerSWClientConnection::retrieveRecordResponseBody):
* Source/WebCore/workers/service/background-fetch/BackgroundFetch.cpp:
(WebCore::BackgroundFetch::updateBackgroundFetchStatus):
(WebCore::BackgroundFetch::unsetRecordsAvailableFlag):
* Source/WebCore/workers/service/background-fetch/BackgroundFetch.h:
(WebCore::BackgroundFetch::options const):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.cpp:
(WebCore::BackgroundFetchEngine::startBackgroundFetch):
(WebCore::BackgroundFetchEngine::notifyBackgroundFetchUpdate):
(WebCore::BackgroundFetchEngine::addFetchFromStore):
(WebCore::BackgroundFetchEngine::backgroundFetch const):
* Source/WebCore/workers/service/background-fetch/BackgroundFetchEngine.h:
* Source/WebCore/workers/service/background-fetch/BackgroundFetchFailureReason.idl:
* Source/WebCore/workers/service/background-fetch/BackgroundFetchResult.idl:
* Source/WebCore/workers/service/context/ServiceWorkerThread.cpp:
(WebCore::ServiceWorkerThread::queueTaskToFireBackgroundFetchEvent):
* Source/WebCore/workers/service/server/SWServer.cpp:
(WebCore::SWServer::fireBackgroundFetchEvent):
* Source/WebCore/workers/service/server/SWServer.h:
* Source/WebKit/CMakeLists.txt:
* Source/WebKit/DerivedSources-input.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/NetworkProcess/NetworkProcess.cpp:
(WebKit::NetworkProcess::getBackgroundFetchState):
* Source/WebKit/NetworkProcess/NetworkProcess.h:
* Source/WebKit/NetworkProcess/NetworkProcess.messages.in:
* Source/WebKit/NetworkProcess/NetworkSession.cpp:
(WebKit::NetworkSession::getBackgroundFetchState):
* Source/WebKit/NetworkProcess/NetworkSession.h:
* Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreImpl.cpp:
(WebKit::BackgroundFetchStoreImpl::clearFetch):
(WebKit::BackgroundFetchStoreImpl::clearAllFetches):
(WebKit::BackgroundFetchStoreImpl::storeFetch):
(WebKit::BackgroundFetchStoreImpl::storeFetchResponseBodyChunk):
(WebKit::BackgroundFetchStoreImpl::getBackgroundFetchState):
* Source/WebKit/NetworkProcess/storage/BackgroundFetchStoreImpl.h:
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.cpp:
(WebKit::NetworkStorageManager::notifyBackgroundFetchChange):
* Source/WebKit/NetworkProcess/storage/NetworkStorageManager.h:
* Source/WebKit/Shared/BackgroundFetchChange.h: Added.
* Source/WebKit/Shared/BackgroundFetchState.h: Added.
* Source/WebKit/Shared/BackgroundFetchState.serialization.in: Added.
* Source/WebKit/Shared/Cocoa/BackgroundFetchStateCocoa.mm: Added.
(WebKit::BackgroundFetchState::toDictionary const):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStore.mm:
(-[WKWebsiteDataStore _getBackgroundFetchState:completionHandler:]):
* Source/WebKit/UIProcess/API/Cocoa/WKWebsiteDataStorePrivate.h:
* Source/WebKit/UIProcess/API/Cocoa/_WKWebsiteDataStoreDelegate.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.cpp:
(WebKit::NetworkProcessProxy::getAllBackgroundFetchIdentifiers):
(WebKit::NetworkProcessProxy::getBackgroundFetchState):
(WebKit::NetworkProcessProxy::notifyBackgroundFetchChange):
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.h:
* Source/WebKit/UIProcess/Network/NetworkProcessProxy.messages.in:
* Source/WebKit/UIProcess/WebsiteData/WebsiteDataStoreClient.h:
(WebKit::WebsiteDataStoreClient::notifyBackgroundFetchChange):
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:
* Tools/WebKitTestRunner/InjectedBundle/Bindings/TestRunner.idl:
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::InjectedBundle::lastAddedBackgroundFetchIdentifier const):
(WTR::InjectedBundle::lastRemovedBackgroundFetchIdentifier const):
(WTR::InjectedBundle::lastUpdatedBackgroundFetchIdentifier const):
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.h:
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.cpp:
(WTR::TestRunner::lastAddedBackgroundFetchIdentifier const):
(WTR::TestRunner::lastRemovedBackgroundFetchIdentifier const):
(WTR::TestRunner::lastUpdatedBackgroundFetchIdentifier const):
* Tools/WebKitTestRunner/InjectedBundle/TestRunner.h:
* Tools/WebKitTestRunner/TestController.cpp:
(WTR::TestController::lastAddedBackgroundFetchIdentifier const):
(WTR::TestController::lastRemovedBackgroundFetchIdentifier const):
(WTR::TestController::lastUpdatedBackgroundFetchIdentifier const):
* Tools/WebKitTestRunner/TestController.h:
* Tools/WebKitTestRunner/TestInvocation.cpp:
(WTR::TestInvocation::didReceiveSynchronousMessageFromInjectedBundle):
* Tools/WebKitTestRunner/cocoa/TestControllerCocoa.mm:
(WTR::TestController::lastAddedBackgroundFetchIdentifier const):
(WTR::TestController::lastRemovedBackgroundFetchIdentifier const):
(WTR::TestController::lastUpdatedBackgroundFetchIdentifier const):
* Tools/WebKitTestRunner/cocoa/TestWebsiteDataStoreDelegate.h:
* Tools/WebKitTestRunner/cocoa/TestWebsiteDataStoreDelegate.mm:
(-[TestWebsiteDataStoreDelegate requestStorageSpace:frameOrigin:quota:currentSize:spaceRequired:decisionHandler:]):
(-[TestWebsiteDataStoreDelegate websiteDataStore:openWindow:fromServiceWorkerOrigin:completionHandler:]):
(-[TestWebsiteDataStoreDelegate notifyBackgroundFetchChange:change:]):
(-[TestWebsiteDataStoreDelegate lastAddedBackgroundFetchIdentifier]):
(-[TestWebsiteDataStoreDelegate lastRemovedBackgroundFetchIdentifier]):
(-[TestWebsiteDataStoreDelegate lastUpdatedBackgroundFetchIdentifier]):

Canonical link: https://commits.webkit.org/261261@main
